### ISSUE TITLE
Training Stops After Epoch 0

### ISSUE THREAD

**Issue by ulliit** [fairy-stockfish/variant-nnue-pytorch#26]
Hi, I am trying to get the NNUE trainer to work on an Ubuntu 20.04 cloud computer. To exclude variant specific issues, I went for normal chess first. 

However, the training stops always after epoch 0 without saving any checkpoint in the specified directory "variant-nnue-pytorch-master/save_230724/default/version_0/checkpoints". Any idea what's going on here?

(env) us@nvdiat4-1vcpu-3-75gb-15cent-ubuntu2004:~/variant-nnue-pytorch-master$ python3.6 train.py --default_root_dir "save_230724" --threads 1 --num-workers 1 --gpus 1 --max_epochs 10 chess_depth5_100mio.bin chess_depth5_100mio.bin
Feature set: HalfKAv2^
Num real features: 45056
Num virtual features: 768
Num features: 45824
Training with chess_depth5_100mio.bin validating with chess_depth5_100mio.bin
Global seed set to 42
Seed 42
Using batch size 16384
Smart fen skipping: True
Random fen skipping: 3
limiting torch to 1 threads.
Using log dir save_230724
/home/us/env/lib/python3.6/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:488: LightningDeprecationWarning: Argument `period` in `ModelCheckpoint` is deprecated in v1.3 and will be removed in v1.5. Please use `every_n_epochs` instead.
  "Argument `period` in `ModelCheckpoint` is deprecated in v1.3 and will be removed in v1.5."
/home/us/env/lib/python3.6/site-packages/pytorch_lightning/callbacks/model_checkpoint.py:433: UserWarning: ModelCheckpoint(save_last=True, save_top_k=None, monitor=None) is a redundant configuration. You can save the last checkpoint with ModelCheckpoint(save_top_k=None, monitor=None).
  "ModelCheckpoint(save_last=True, save_top_k=None, monitor=None) is a redundant configuration."
ModelCheckpoint(save_last=True, save_top_k=-1, monitor=None) will duplicate the last checkpoint saved.
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
Using c++ data loader
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Ranger optimizer loaded. 
Gradient Centralization usage = False

  | Name         | Type                          | Params

0 | input        | DoubleFeatureTransformerSlice | 23.8 M
1 | layer_stacks | LayerStacks                   | 152 K 

24.0 M    Trainable params
0         Non-trainable params
24.0 M    Total params
95.925    Total estimated model params size (MB)
Global seed set to 42                                                                                                                                                                                      
Epoch 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1283/1283 [02:02<00:00, 10.44it/s, loss=0.0129, v_num=0Killed

--------------------------------------

This is my installed python environment:

(env) us@nvdiat4-1vcpu-3-75gb-15cent-ubuntu2004:~$ python3.6 -m pip freeze
absl-py==1.4.0
aiohttp==3.8.4
aiosignal==1.2.0
async-timeout==4.0.2
asynctest==0.13.0
attrs==22.2.0
cachetools==4.2.4
certifi==2023.5.7
charset-normalizer==2.0.12
cmake==3.26.4
cupy-cuda101==9.6.0
cycler==0.11.0
dataclasses==0.8
fastrlock==0.8.1
frozenlist==1.2.0
fsspec==2022.1.0
future==0.18.3
google-auth==2.22.0
google-auth-oauthlib==0.4.6
grpcio==1.48.2
idna==3.4
idna-ssl==1.1.0
importlib-metadata==4.8.3
importlib-resources==5.4.0
install==1.3.5
kiwisolver==1.3.1
Markdown==3.3.7
matplotlib==3.3.4
multidict==5.2.0
numpy==1.19.5
oauthlib==3.2.2
packaging==21.3
Pillow==8.4.0
protobuf==3.19.6
pyasn1==0.5.0
pyasn1-modules==0.3.0
pyDeprecate==0.3.1
pyparsing==3.1.0
python-chess==0.31.4
python-dateutil==2.8.2
pytorch-lightning==1.4.9
PyYAML==6.0
requests==2.27.1
requests-oauthlib==1.3.1
rsa==4.9
six==1.16.0
tensorboard==2.10.1
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
torch==1.8.1+cu101
torchmetrics==0.7.3
tqdm==4.64.1
typing_extensions==4.1.1
urllib3==1.26.16
Werkzeug==2.0.3
yarl==1.7.2
zipp==3.6.0

(env) us@nvdiat4-1vcpu-3-75gb-15cent-ubuntu2004:~$ python3.6
Python 3.6.15 (default, Apr 25 2022, 01:55:53) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.version.cuda
'10.1'

(env) us@nvdiat4-1vcpu-3-75gb-15cent-ubuntu2004:~$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243

---------------------------------------------------

The training data file was generated on another, cpu-only cloud machine:

us@c2-16vcpu-120gb-ubuntu2004-9cent:~$ ./fairy-stockfish-tools_x86-64-bmi2
Fairy-Stockfish 260422 by Fabian Fichter
uci
...
uciok
setoption name Use NNUE value false
setoption name Threads value 8
setoption name Hash value 2048
setoption name UCI_Variant value chess
lib/nnue_training_data_formats.h:
#define FILES 8
#define RANKS 8
#define PIECE_TYPES 6
#define PIECE_COUNT 32
#define POCKETS false
#define KING_SQUARES 64

variant.py:
RANKS = 8
FILES = 8
SQUARES = RANKS * FILES
KING_SQUARES = 64
PIECE_TYPES = 6
PIECES = 2 * PIECE_TYPES
USE_POCKETS = False
POCKETS = 2 * FILES if USE_POCKETS else 0

PIECE_VALUES = {
  1: 126,
  2: 781,
  3: 825,
  4: 1276,
  5: 2538,
}
info string variant chess files 8 ranks 8 pocket 0 template fairy startpos rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
isready
readyok
generate_training_data depth 5 count 100000000 random_multi_pv 4 random_multi_pv_diff 100 random_move_count 8 random_move_max_ply 20 write_min_ply 5 eval_limit 10000 set_recommended_uci_options data_format bin output_file_name chess_depth5_100mio.bin
INFO: Executing generate_training_data command
INFO: Parameters:
  - search_depth_min       = 5
  - search_depth_max       = 5
  - nodes                  = 0
  - count                  = 100000000
  - eval_limit             = 10000
  - eval_diff_limit        = 500
  - num threads (UCI)      = 8
  - random_move_min_ply    = 1
  - random_move_max_ply    = 20
  - random_move_count      = 8
  - random_move_like_apery = 0
  - random_multi_pv        = 4
  - random_multi_pv_diff   = 100
  - random_multi_pv_depth  = 5
  - write_min_ply          = 5
  - write_max_ply          = 400
  - book                   = 
  - output_file_name       = chess_depth5_100mio.bin
  - save_every             = 18446744073709551615
  - random_file_name       = 0
  - write_drawn_games      = 1
  - draw by low score      = 1
  - draw by insuff. mat.   = 1
info string classical evaluation enabled
INFO (sfen_writer): Creating new data file at chess_depth5_100mio.bin
PRNG::seed = cbbe2308f46191ed
........................................
200000 sfens, 12277 sfens/second, at Sat Jul 22 03:47:07 2023
...
100000000 sfens, 10813 sfens/second, at Sat Jul 22 06:18:04 2023

INFO: generate_training_data finished.
quit
us@c2-16vcpu-120gb-ubuntu2004-9cent:~$

**Comment by ianfab** [comment]
At first glance to me the most likely should be a lack of write permissions in the output directory.

If there was anything wrong with the training data, the packages, the GPU, or the like, the training would very likely already fail in the training or show a weird loss, but that looks all fine. However, what I find a bit weird is that it shows "killed" and does not print any stack trace of the failed python code. This might possibly indicate an OS intervention, e.g., due to running out of RAM or disk space.

**Comment by ulliit** [comment]
Indeed, increasing the RAM did the job 👍️👍️👍️. Thank you very much!
